File: 1-working-with-list/p03_at.ml

let rec at n l =
  match l with
  | [] -> None
  | hd :: tail when n > 0 -> if n = 1 then Some hd else at (n-1) tail
  | _ -> None ;;

at 3 [ "a" ; "b"; "c"; "d"; "e" ];;
at 3 [ "a" ; "b"; "c"];;
at 3 [ "a" ];;
at 1 [ "a" ];;
at 3 [ ];;
at (-3) [ "a" ; "b"; "c"];;

let rec at' n l =
  if n < 1
  then None
  else
    match l with
    | [] -> None
    | hd :: tl -> if n = 1 then Some hd else at (n-1) tl;;

at' 3 [ "a" ; "b"; "c"; "d"; "e" ];;
at' 3 [ "a" ; "b"; "c"];;
at' 3 [ "a" ];;
at' 1 [ "a" ];;
at' 3 [ ];;
at' (-3) [ "a" ; "b"; "c"];;

File: 1-working-with-list/p03_at.scala

object Main {

  def main(args: Array[String]) {
    println(at(3, List("a", "b", "c", "d")))
  }

  def at[T](idx: Int, l: List[T]): Option[T] = {
    if(idx > l.length || idx <= 0)
      None
    else {
      l match {
        case Nil => None
        case hd :: rest => if(idx == 1) Some(hd) else at(idx-1, rest)
      }
    }
  }
}
